src/engraving/dom/undo.cpp

```diff
@@ -854,6 +854,11 @@ void AddElement::endUndoRedo(bool isUndo) const
     } else if (element->isKeySig()) {
         element->triggerLayout();
         element->score()->setLayout(element->staff()->nextKeyTick(element->tick()), element->staffIdx());
+    } else if (element->isHarpPedalDiagram()) {
+        HarpPedalDiagram* nextHd = element->part()->nextHarpDiagram(element->tick());
+        Fraction nextHdTick = nextHd ? nextHd->tick() : element->score()->endTick();
+        element->score()->setLayout(element->tick(), nextHdTick, element->part()->staves().front()->idx(),
+                                    element->part()->staves().back()->idx());
```

Comment:
Why do we need to re-layout the score when a diagram changes? 

If it is because the next text diagram might need to be updated: don't the diagrams _after_ that need to be updated too?

Also, I would expect that we _do_ need to update the playback model; or is that working already? 
Anyway, updating the playback model is done based on the "object items" returned by the `objectItems()` method of `UndoCommand`. So, it looks like for example `ChangeHarpPedalState::objectItems()` will need to return a list containing all affected glissandi. I'm not sure if that's ideal... As a hack, you could add just the next pedal diagram to the list; that will work because the playback model works based on a tick range generated from the list of objectItems. 
A better solution would be to refactor the whole `objectItems` thing, and return only the _directly_ affected items there, and let the playback model figure out for which other items playback data needs to be updated. That way, we keep the `objectItems` general-purpose, as they should, instead of biased towards playback (which doesn't make sense, since general undo/redo logic should not be aware of playback logic).

Reply:
Yes, this was an attempt to update the playback model.  Previous to this I had a solution which, like you suggested, returned the next pedal diagram from `objectItems()`, and if there were no following pedal diagrams, the segment at the end of the score to make sure all following glissandos were updated.  I think I changed it as it felt a bit hacky, but is it potentially a better solution than the current one?

Reply:
The thing is, I don't see how triggering a re-layout triggers an update of the playback model... it's very possible that I'm just missing it though, but as far as I can see the PlaybackModel only listens to the changesChannel of the score, and that is completely disconnected from triggering re-layouts.

Reply:
So as far as I worked out - `score()->setLayout()` sets the start and end ticks for `CmdState`, which are used to determine the start and end ticks of `Score::changesRange()`. This is sent to `changesChannel` on `Score::endCmd`, which updates the playback model.  I agree, not a very clear way for me to do it...

Reply:
An alternative - remove these layout changes and add this:
```cpp
std::vector<const EngravingObject*> ChangeHarpPedalState::objectItems() const
{
    Part* part = diagram->part();
    std::vector<const EngravingObject*> objs{ diagram };
    if (!part) {
        return objs;
    }

    HarpPedalDiagram* nextDiagram = part->nextHarpDiagram(diagram->tick());
    if (nextDiagram) {
        objs.push_back(nextDiagram);
    } else {
        objs.push_back(diagram->score()->lastElement());
    }
    return objs;
}
```

Reply:
I see. I think I prefer the alternative, because redoing layout of a big part of the score that hasn't really changed may be bad for performance. And I think the hack via objectItems is not much hackier than a hack via CmdState.